pow/src/main/java/tech/pegasys/teku/pow/MinimumGenesisTimeBlockFinder.java

```diff
@@ -52,7 +60,93 @@ public MinimumGenesisTimeBlockFinder(
       final BigInteger headBlockNumber) {
     return binarySearchLoop(
-        new SearchContext(eth1DepositContractDeployBlock.orElse(ZERO), headBlockNumber));
+            new SearchContext(eth1DepositContractDeployBlock.orElse(ZERO), headBlockNumber))
+        .thenCompose(this::confirmOrFindMinGenesisBlock);
+  }
+
+  /**
+   * The binary search may return a block that is too new (if historical blocks are unavailable
+   * during the search), so walk back though chain to confirm or else pull the correct min genesis
+   * block.
+   *
+   * @param candidateMinGenesisBlock A candidate block that may be the min genesis block
+   * @return The confirmed min genesis block
+   */
+  @VisibleForTesting
+  SafeFuture<EthBlock.Block> confirmOrFindMinGenesisBlock(
+      final EthBlock.Block candidateMinGenesisBlock) {
+    assertBlockIsAtOrAfterMinGenesis(candidateMinGenesisBlock);
+    if (blockIsAtMinimalMinGenesisPosition(candidateMinGenesisBlock)) {
+      // We've found min genesis
+      traceWithBlock(
+          "Confirmed minimum genesis block at minimal position: ", candidateMinGenesisBlock);
+      return SafeFuture.completedFuture(candidateMinGenesisBlock);
+    }
+
+    traceWithBlock(
+        "Confirm minimum genesis block retrieved via binary search: ", candidateMinGenesisBlock);
+    final SafeFuture<EthBlock.Block> minGenesisFuture = new SafeFuture<>();
+
+    final AtomicReference<EthBlock.Block> currentMinGenesisCandidate =
+        new AtomicReference<>(candidateMinGenesisBlock);
+    final AtomicInteger examinedAncestorsCount = new AtomicInteger(0);
+    SafeFuture.asyncDoWhile(
+            () ->
+                eth1Provider
+                    .getEth1BlockWithRetry(
+                        currentMinGenesisCandidate.get().getParentHash(), Duration.ofSeconds(5), 3)
+                    .thenApply(
+                        parent -> {
+                          examinedAncestorsCount.incrementAndGet();
+                          if (blockIsPriorToMinGenesis(parent)) {
+                            // We've confirmed the current candidate is at the boundary
+                            traceWithBlock(
+                                "Confirmed min genesis block: ", currentMinGenesisCandidate.get());
+                            minGenesisFuture.complete(currentMinGenesisCandidate.get());
+                            return false;
+                          } else if (blockIsAtMinGenesis(parent)) {
+                            // We've found the min genesis block
+                            traceWithBlock(
+                                "Confirmed min genesis block at exact genesis time: ", parent);
+                            minGenesisFuture.complete(parent);
+                            return false;
+                          } else {
+                            // The parent block is after min genesis
+                            // Keep searching through ancestors up to our limit
```

Comment:
This is the approach we talked about using - but I'm not sure it actually makes a lot of sense to search one by one through the ancestors.  I suppose it's possible that some historical blocks were filled in as the binary search runs, but it's also possible we spend a lot of time walking back through ancestors and don't find our target block.  

Thoughts?

Reply:
Yeah it is an awful lot of extra code for a very small corner case.  It only matters if we don't already have the genesis state baked in and the eth1 node doesn't have enough history to get back to the min genesis block but then it wouldn't have the deposit events anyway and we'd crash because of the deposit event ordering checks.

I think its probably worth just dropping this - keep the assumption that if we can't get a block then min genesis time must be after that but don't bother walking back to verify it was right.

Reply:
I kept a single check to look up the parent block, but dropped the ancestor search which cut out a good chunk of code.